field_audit/models.py

```diff
@@ -0,0 +1,259 @@
+    is_create = models.BooleanField(default=False)
+    is_delete = models.BooleanField(default=False)
+    delta = models.JSONField()
+
+    objects = get_manager("AUDITEVENT_MANAGER", DefaultAuditEventManager)
+
+    class Meta:
+        indexes = [
+            GinIndex(fields=["delta"]),
+        ]
+        constraints = [
+            models.CheckConstraint(
+                name="field_audit_auditevent_valid_create_or_delete",
+                check=~models.Q(is_create=True, is_delete=True),
+            ),
+        ]
+
+    ATTACH_INIT_VALUES_AT = "__field_audit_init_values"
+
+    @staticmethod
+    def get_field_value(instance, field_name):
+        """Returns the database value of a field on ``instance``.
+
+        :param instance: an instance of a Django model
+        :param field_name: name of a field on ``instance``
+        """
+        return instance._meta.get_field(field_name).value_from_object(instance)
+
+    @classmethod
+    def attach_initial_values(cls, field_names, instance):
+        """Save copies of field values on an instance so they can be used later
+        to determine if the instance has changed and record what the previous
+        values were.
+
+        :param field_names: a collection of names of fields on ``instance``
+            to attach for later auditing
+        :param instance: instance of a Model subclass to be audited for changes
+        :raises: ``AttachValuesError`` if initial values are already attached to
+            the instance
+        """
+        if hasattr(instance, cls.ATTACH_INIT_VALUES_AT):
+            # This should never happen, but to be safe, refuse to clobber
+            # existing attributes.
+            raise AttachValuesError(
+                f"refusing to overwrite {cls.ATTACH_INIT_VALUES_AT!r} "
+                f"on model instance: {instance}"
+            )
+        init_values = {f: cls.get_field_value(instance, f) for f in field_names}
+        setattr(instance, cls.ATTACH_INIT_VALUES_AT, init_values)
+
+    @classmethod
+    def reset_initial_values(cls, field_names, instance):
+        """Returns the previously attached "initial values" and attaches new
+        values.
+
+        :param field_names: a collection of names of fields on ``instance``
+            attach for later auditing
+        :param instance: instance of a Model subclass to be audited for changes
+        :raises: ``AttachValuesError`` if initial values are not attached to
+            the instance
+        """
+        try:
+            values = getattr(instance, cls.ATTACH_INIT_VALUES_AT)
+        except AttributeError:
+            raise AttachValuesError("cannot reset values that were never set")
+        delattr(instance, cls.ATTACH_INIT_VALUES_AT)
+        cls.attach_initial_values(field_names, instance)
+        return values
+
+    @classmethod
+    def audit_field_changes(cls, field_names, instance, is_create, is_delete,
+                            request, object_pk=None):
```

Comment:
Should `request` be optional? Seems like you'd want to be able to audit changes that happen outside of a request e.g. shell / management command. At least the docstring should note that it can be `None`